Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1

```diff
@@ -150,18 +169,49 @@ function Set-CurrentAzureRMSubscription {
     $null = Select-AzureRMSubscription -SubscriptionId $SubscriptionId @additional
 }
 
-function Format-Splat {
+function Set-UserAgent {
     [CmdletBinding()]
-    param([Parameter(Mandatory = $true)][hashtable]$Hashtable)
-
-    # Collect the parameters (names and values) in an array.
-    $parameters = foreach ($key in $Hashtable.Keys) {
-        $value = $Hashtable[$key]
-        # If the value is a bool, format the parameter as a switch (ending with ':').
-        if ($value -is [bool]) { "-$($key):" } else { "-$key" }
-        $value
+    param()
+
+    $collectionUri = Get-VstsTaskVariable -Name System.TeamFoundationCollectionUri -Require
+    $collectionId = Get-VstsTaskVariable -Name System.CollectionId -Require
+    $hostType = Get-VstsTaskVariable -Name System.HostType -Require
+    $serverString = "TFS"
+    if ($collectionUri.ToLower().Contains("visualstudio.com".ToLower())) {
+        $serverString = "VSTS"
     }
+    
+    $userAgent = [string]::Empty
+    if ($hostType -ieq "build") {
+		$definitionId = Get-VstsTaskVariable -Name System.DefinitionId -Require
+		$buildId = Get-VstsTaskVariable -Name Build.BuildId -Require
+		$userAgent = $serverString + "_" + $collectionId + "_" + "build" + "_" + $definitionId + "_" + $buildId
+    } elseif ($hostType -ieq "release") {
+        $definitionId = Get-VstsTaskVariable -Name Release.DefinitionId -Require
+		$releaseId = Get-VstsTaskVariable -Name Release.ReleaseId -Require
+		$environmentId = Get-VstsTaskVariable -Name Release.EnvironmentId -Require
+		$attemptNumber = Get-VstsTaskVariable -Name Release.AttemptNumber -Require
+		$userAgent = $serverString + "_" + $collectionId + "_" + "release" + "_" + $definitionId + "_" + $releaseId + "_" + $environmentId + "_" + $attemptNumber
+    } else {
+	    return
+	}
+	
+    Set-UserAgent_Core -UserAgent $userAgent
+}
 
-    $OFS = " "
-    "$parameters" # String join the array.
+function Set-UserAgent_Core {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [string]$UserAgent)
+
+    Trace-VstsEnteringInvocation $MyInvocation
+    try {
+        [Microsoft.Azure.Common.Authentication.AzureSession]::ClientFactory.AddUserAgent($UserAgent)
```

Comment:
* Does this work with all of the range of Azure/AzureRM module versions currently [supported by this module](https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/Common/VstsAzureHelpers_/ImportFunctions.ps1#L39)? It might be good to add some comments for anyone else looking at the source code what the range of versions you tested against.
* Regarding whether this works if AzureRM and Azure are both imported. There likely is a scenario where it wouldn't if different versions of the underlying DLL are loaded (where the type lives). I'm not sure what a better solution would be.